src/main/java/org/folio/services/locationunit/LibraryService.java

```diff
@@ -0,0 +1,153 @@
+package org.folio.services.locationunit;
+
+import static org.folio.rest.impl.StorageHelper.logAndSaveError;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Context;
+import io.vertx.core.Future;
+import io.vertx.sqlclient.Row;
+import io.vertx.sqlclient.RowSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+import javax.ws.rs.core.Response;
+import org.folio.persist.LibraryRepository;
+import org.folio.rest.jaxrs.model.Error;
+import org.folio.rest.jaxrs.model.Errors;
+import org.folio.rest.jaxrs.model.Loclib;
+import org.folio.rest.jaxrs.model.Loclibs;
+import org.folio.rest.jaxrs.model.Parameter;
+import org.folio.rest.jaxrs.resource.LocationUnits.DeleteLocationUnitsLibrariesByIdResponse;
+import org.folio.rest.jaxrs.resource.LocationUnits.DeleteLocationUnitsLibrariesResponse;
+import org.folio.rest.jaxrs.resource.LocationUnits.GetLocationUnitsLibrariesByIdResponse;
+import org.folio.rest.jaxrs.resource.LocationUnits.GetLocationUnitsLibrariesResponse;
+import org.folio.rest.jaxrs.resource.LocationUnits.PostLocationUnitsLibrariesResponse;
+import org.folio.rest.jaxrs.resource.LocationUnits.PutLocationUnitsLibrariesByIdResponse;
+import org.folio.rest.persist.PgUtil;
+import org.folio.services.domainevent.LibraryDomainEventPublisher;
+
+public class LibraryService {
+
+  public static final String LIBRARY_TABLE = "loclibrary";
+
+  private final Context context;
+  private final Map<String, String> okapiHeaders;
+  private final LibraryRepository repository;
+  private final LibraryDomainEventPublisher domainEventService;
+
+  public LibraryService(Context context, Map<String, String> okapiHeaders) {
+    this.context = context;
+    this.okapiHeaders = okapiHeaders;
+    this.repository = new LibraryRepository(context, okapiHeaders);
+    this.domainEventService = new LibraryDomainEventPublisher(context, okapiHeaders);
+  }
+
+  public Future<Response> getByQuery(String cql, int offset, int limit) {
+    return PgUtil.get(LIBRARY_TABLE, Loclib.class, Loclibs.class,
+      cql, offset, limit, okapiHeaders, context, GetLocationUnitsLibrariesResponse.class);
+  }
+
+  public Future<Response> getById(String id) {
+    return PgUtil.getById(LIBRARY_TABLE, Loclib.class, id, okapiHeaders, context,
+      GetLocationUnitsLibrariesByIdResponse.class);
+  }
+
+  public Future<Response> create(Loclib loclib) {
+    return PgUtil.post(LIBRARY_TABLE, loclib, okapiHeaders, context, PostLocationUnitsLibrariesResponse.class)
+      .onSuccess(domainEventService.publishCreated());
+  }
+
+  public Future<Response> update(String id, Loclib loclib) {
+    if (loclib.getId() == null) {
+      loclib.setId(id);
+    }
+    var checks = List.of(checkIdChange(id, loclib));
+    return doLibraryCheck(checks)
```

Comment:
I don't understand why, for a simple check in the service, you need to:

- Create a method doLibraryCheck with a List as a parameter.
- Create a special exception that you then convert into an Errors class object.

It would be better to perform the check and, if it fails, immediately return a completed future with an Errors response

Reply:
First I did based on LocationApi pr, removed this part and do much improvements based on you feedback. Thanks.